examples/parse-body-based-on-content-type/src/main.rs

```diff
@@ -0,0 +1,98 @@
+//! Provides a RESTful web server managing some Todos.
+//!
+//! API will be:
+//!
+//! - `GET /todos`: return a JSON list of Todos.
+//! - `POST /todos`: create a new Todo.
+//! - `PUT /todos/:id`: update a specific Todo.
+//! - `DELETE /todos/:id`: delete a specific Todo.
+//!
+//! Run with
+//!
+//! ```not_rust
+//! cd examples && cargo run -p example-todos
+//! ```
+
+use axum::{
+    async_trait,
+    extract::FromRequest,
+    http::{header::CONTENT_TYPE, Request, StatusCode},
+    response::{IntoResponse, Response},
+    routing::post,
+    Form, Json, RequestExt, Router,
+};
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddr;
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::EnvFilter::new(
+            std::env::var("RUST_LOG").unwrap_or_else(|_| {
+                "example_parse_body_based_on_content_type=debug,tower_http=debug".into()
+            }),
+        ))
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let app = Router::new().route("/", post(handler));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+#[derive(Serialize, Deserialize)]
+struct Payload {
+    foo: String,
+}
+
+async fn handler(payload: JsonOrForm<Payload>) -> Response {
+    match payload {
+        JsonOrForm::Json(payload) => Json(payload).into_response(),
+        JsonOrForm::Form(payload) => Form(payload).into_response(),
+    }
+}
+
+enum JsonOrForm<T, K = T> {
```

Comment:
Why is the second type parameter named `K`?

Reply:
Might be over engineered but the idea was to be able to extract different types for json and form requests.

Reply:
Yeah I got that, the question was just why `T` and `K`, rather than e.g. `T` and `U` or `A` and `B` which are a much more common combination for generic parameters without specific meaning.

Reply:
Ah okay. No particular reason. I'll change it to A and B

Reply:
Actually I'll use T and U because B is taken by the request body.

Reply:
I'll just leave it like this :P